### PR TITLE
Add decompositions for reflection_pad{1, 2, 3}d

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115101
* #115113
* __->__ #115100

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by Chillee** [comment]
I think the performance of this is going to be worse than the current implementation of `reflection_pad`. The current implementation doesn't require a `concat`, while the decomposition does.

**Comment by isuruf** [comment]
@Chillee, thanks for reviewing. The concatenation is on the indices and not on the matrix.

I ran a benchmark with input of size `3,5,1000,1000` for `torch.nn.ReflectionPad2d((2, 3, 4, 5))` and the results did not show any slowdown except for small inputs.

```
[--------------------- reflection_pad2d ---------------------]
                          |  Decomposed  |  Lowering  |  Eager
32 threads: --------------------------------------------------
      (3, 5, 1000, 10)    |      15      |     11     |    12 
      (3, 5, 1000, 100)   |      44      |     54     |    56 
      (3, 5, 1000, 1000)  |     416      |    501     |   504 

Times are in microseconds (us).
 ```
 
Generated triton kernel:

<details>

```python
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty, empty_strided
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
alloc_from_pool = torch.ops.inductor._alloc_from_pool
reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
async_compile = AsyncCompile()


# kernel path: /tmp/torchinductor_isuruf/cs/ccsxzv5jaeadushd7qxpyfwilwnaxwku3cwxjonlbpxbmvflqgpa.py
# Source Nodes: [], Original ATen: []

triton_poi_fused_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import AutotuneHint, pointwise
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@pointwise(
    size_hints=[16777216], 
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [instance_descriptor(divisible_by_16=(0, 1), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_0', 'mutated_arg_names': []},
    min_elem_per_thread=0
)
@triton.jit
def triton_(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 15210675
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex % 1005
    x1 = (xindex // 1005) % 1009
    x2 = (xindex // 1014045)
    x5 = xindex
    tmp0 = x0
    tmp1 = tl.full([1], 0, tl.int64)
    tmp2 = tmp0 >= tmp1
    tmp3 = tl.full([1], 2, tl.int64)
    tmp4 = tmp0 < tmp3
    tmp5 = 2 + ((-1)*x0)
    tmp6 = tl.full(tmp5.shape, 0.0, tmp5.dtype)
    tmp7 = tl.where(tmp4, tmp5, tmp6)
    tmp8 = tmp0 >= tmp3
    tmp9 = tl.full([1], 1002, tl.int64)
    tmp10 = tmp0 < tmp9
    tmp11 = tmp8 & tmp10
    tmp12 = (-2) + x0
    tmp13 = tl.full(tmp12.shape, 0.0, tmp12.dtype)
    tmp14 = tl.where(tmp11, tmp12, tmp13)
    tmp15 = tmp0 >= tmp9
    tmp16 = tl.full([1], 1005, tl.int64)
    tmp17 = tmp0 < tmp16
    tmp18 = 2000 + ((-1)*x0)
    tmp19 = tl.full(tmp18.shape, 0.0, tmp18.dtype)
    tmp20 = tl.where(tmp15, tmp18, tmp19)
    tmp21 = tl.where(tmp11, tmp14, tmp20)
    tmp22 = tl.where(tmp4, tmp7, tmp21)
    tmp23 = tmp22 + 1000
    tmp24 = tmp22 < 0
    tmp25 = tl.where(tmp24, tmp23, tmp22)
    tl.device_assert((0 <= tmp25) & (tmp25 < 1000), "index out of bounds: 0 <= tmp25 < 1000")
    tmp26 = x1
    tmp27 = tmp26 >= tmp1
    tmp28 = tl.full([1], 4, tl.int64)
    tmp29 = tmp26 < tmp28
    tmp30 = 4 + ((-1)*x1)
    tmp31 = tl.full(tmp30.shape, 0.0, tmp30.dtype)
    tmp32 = tl.where(tmp29, tmp30, tmp31)
    tmp33 = tmp26 >= tmp28
    tmp34 = tl.full([1], 1004, tl.int64)
    tmp35 = tmp26 < tmp34
    tmp36 = tmp33 & tmp35
    tmp37 = (-4) + x1
    tmp38 = tl.full(tmp37.shape, 0.0, tmp37.dtype)
    tmp39 = tl.where(tmp36, tmp37, tmp38)
    tmp40 = tmp26 >= tmp34
    tmp41 = tl.full([1], 1009, tl.int64)
    tmp42 = tmp26 < tmp41
    tmp43 = 2002 + ((-1)*x1)
    tmp44 = tl.full(tmp43.shape, 0.0, tmp43.dtype)
    tmp45 = tl.where(tmp40, tmp43, tmp44)
    tmp46 = tl.where(tmp36, tmp39, tmp45)
    tmp47 = tl.where(tmp29, tmp32, tmp46)
    tmp48 = tmp47 + 1000
    tmp49 = tmp47 < 0
    tmp50 = tl.where(tmp49, tmp48, tmp47)
    tl.device_assert((0 <= tmp50) & (tmp50 < 1000), "index out of bounds: 0 <= tmp50 < 1000")
    tmp51 = tl.load(in_ptr0 + (tmp25 + (1000*tmp50) + (1000000*x2)), xmask, eviction_policy='evict_last')
    tl.store(out_ptr0 + (x5), tmp51, xmask)
''')

import triton
import triton.language as tl
from torch._inductor.triton_heuristics import grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_cuda_stream


async_compile.wait(globals())
del async_compile

def call(args):
    args_1, = args
    args.clear()
    assert_size_stride(args_1, (3, 5, 1000, 1000), (5000000, 1000000, 1000, 1))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty((3, 5, 1009, 1005), device='cuda', dtype=torch.float32)
        # Source Nodes: [], Original ATen: []
        stream0 = get_cuda_stream(0)
        triton_poi_fused_0.run(args_1, buf0, 15210675, grid=grid(15210675), stream=stream0)
        del args_1
        return (buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    args_1 = rand_strided((3, 5, 1000, 1000), (5000000, 1000000, 1000, 1), device='cuda:0', dtype=torch.float32)
    return print_performance(lambda: call([args_1]), times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
``` 

</details> 

Benchmark details

<details>

 ```python
 import torch
from torch.testing import make_tensor
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.benchmark import Timer, Compare
from torch._inductor.compile_fx import compile_fx_inner, cudagraphify_impl
from torch._inductor.decomposition import decompositions
from itertools import product
from functools import partial

torch._logging.set_logs(output_code=True)

benchmark_name = "reflection_pad2d"
Ss = [[3, 5, 1000, 10], [3, 5, 1000, 100], [3, 5, 1000, 1000]]


def gen_inputs():
    for shape in Ss:
        yield torch.randn(shape, dtype=torch.float32, device="cuda")

def benchmark(label, f, x):
    return Timer("f([x])",
                 globals=locals(),
                 label=benchmark_name,
                 description=label,
                 sub_label=f"{tuple(x.shape)}",
                 num_threads=torch.get_num_threads()).blocked_autorange(min_run_time=2)

def compare(x):
    def f(args):
        x, = args
        val = torch.nn.ReflectionPad2d((2, 3, 4, 5))(x)
        return (val,)

    print(f"{tuple(x.shape)}")
    args = [x]

    decomposed = make_fx(f, decomposition_table=decompositions, tracing_mode="fake")(args)
    compiled_decomposed = compile_fx_inner(decomposed, args, cudagraphs=False)
    yield benchmark("Decomposed", compiled_decomposed, *args)

    non_decomposed = make_fx(f, tracing_mode="fake")(args)
    compiled_nondecomposed = compile_fx_inner(non_decomposed, args, cudagraphs=False)
    yield benchmark("Lowering", compiled_nondecomposed, *args)

    # Just show the first two generated kernels
    torch._logging.set_logs(output_code=False)

    cuda_f = cudagraphify_impl(f, args, static_input_idxs=tuple(range(len(args))))
    yield benchmark("Eager", cuda_f, *args)

results = []
for x in gen_inputs():
    for res in compare(x):
        results.append(res)

compare = Compare(results)
compare.trim_significant_figures()
compare.print()
```
</details>

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -4114,6 +4114,53 @@ def upsample_bicubic2d_vec(
...
+        else:
+            right = torch.arange(
+                inp_shape[i] - 2,
+                inp_shape[i] - 2 - padding_right[i],
+                step=-1,
+                device=a.device,
+            )
+            to_cat.append(right)
+        idx: List[Any] = [slice(s) for s in result.shape]
+        idx[i + nc_dim] = torch.cat(tuple(to_cat))
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:4155]: Could we avoid `cat` by doing the same index calculation used in the lowering
```python
dim_idx = torch.arange(-padding_left[i], inp_shape[i] + padding_right[i], device=a.device)
dim_idx = inp_shape[i] - (inp_shape[i] - dim_idx.abs()).abs()
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -4114,6 +4114,53 @@ def upsample_bicubic2d_vec(
...
+        result = result[idx]
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:4156]: We should use `_unsafe_index` here which skips bounds checking, same as `check=False` which is used in the lowering.
```suggestion
        result = aten._unsafe_index(result, idx)
```

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (isuruf) [reply torch/_decomp/decompositions.py]: Thanks. That makes the implementation and the generated code much nicer.

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (isuruf) [reply torch/_decomp/decompositions.py]: Thanks

**Review by peterbell10 — APPROVED** [review]

**Comment by isuruf** [comment]
@Chillee, I've removed the concat implementation thanks to @peterbell10's suggestion.

**Comment by isuruf** [comment]
I forgot to add the benchmarks here.
```
[------------------ reflect_pad -------------------]
                             |  Decomposed  |  Eager
32 threads: ----------------------------------------
      (3, 5, 1000000)        |    432.4     |   416 
      (3, 5, 1000, 1000)     |    433.7     |   500 
      (3, 5, 100, 100, 100)  |    446.4     |   808 
```

**Pull request closed without merging**